kubectl/upload/main.go

```diff
@@ -0,0 +1,212 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type Config struct {
+	Verbose    bool
+	Path       string
+	Kubeconfig string
+	Resource
+}
+
+type Resource struct {
+	Kind       string
+	Name       string
+	Namespace  string
+	EncodedMd5 string
+	// TODO(bjb): clean this up when the CRD changes we should only pass around
+	// encodedMd5s and not hex checksums
+	Md5Checksum string
+}
+
+type KubernetesClient interface {
+	CreateResource(gvr schema.GroupVersionResource, namespace string, obj *unstructured.Unstructured, opts metav1.CreateOptions) (*unstructured.Unstructured, error)
+	WatchResource(gvr schema.GroupVersionResource, namespace string, opts metav1.ListOptions) (watch.Interface, error)
+}
+
+type realKubernetesClient struct {
+	dynamicClient dynamic.Interface
+}
+
+func (c *realKubernetesClient) CreateResource(gvr schema.GroupVersionResource, namespace string, obj *unstructured.Unstructured, opts metav1.CreateOptions) (*unstructured.Unstructured, error) {
+	return c.dynamicClient.Resource(gvr).Namespace(namespace).Create(context.Background(), obj, opts)
+}
+
+func (c *realKubernetesClient) WatchResource(gvr schema.GroupVersionResource, namespace string, opts metav1.ListOptions) (watch.Interface, error) {
+	return c.dynamicClient.Resource(gvr).Namespace(namespace).Watch(context.Background(), opts)
+}
+
+func main() {
+	var cfg Config
+
+	var cmd = &cobra.Command{
+		Use:   "upload [kind] [name]",
+		Short: "Upload a resource of a given Kind and Name to be built as a substratus container image",
+		Args:  cobra.ExactArgs(2),
+		Run: func(cmd *cobra.Command, args []string) {
+			cfg.Resource.Kind = args[0]
+			cfg.Resource.Name = args[1]
+
+			config, _ := clientcmd.BuildConfigFromFlags("", cfg.Kubeconfig)
+			dynamicClient, err := dynamic.NewForConfig(config)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			client := &realKubernetesClient{
+				dynamicClient: dynamicClient,
+			}
+
+			err = run(cfg, client)
+			if err != nil {
+				log.Fatal(err)
+			}
+		},
+	}
+	cmd.Flags().StringVarP(&cfg.Path, "path", "p", ".", "Path to the directory to be uploaded")
+	cmd.Flags().StringVarP(&cfg.Resource.Namespace, "namespace", "n", "default", "Namespace of the resource created")
+	if home := homeDir(); home != "" {
+		cmd.Flags().StringVarP(&cfg.Kubeconfig, "kubeconfig", "", filepath.Join(home, ".kube", "config"), "")
+	} else {
+		cmd.Flags().StringVarP(&cfg.Kubeconfig, "kubeconfig", "", "", "")
+	}
+	cmd.Flags().BoolVarP(&cfg.Verbose, "verbose", "v", false, "Verbose output")
+
+	if err := cmd.Execute(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run(cfg Config, client KubernetesClient) error {
+	if !fileExists(filepath.Join(cfg.Path, "Dockerfile")) {
+		return errors.New("a Dockerfile does not exist at the given path")
+	}
+
+	if cfg.Verbose {
+		fmt.Println("packaging the directory into a tarball")
+	}
+
+	tarPath := "/tmp/archive.tar.gz"
```

Comment:
Hardcoded this but open to suggestions.

Reply:
Tar to `bytes.Buffer` (i.e. in mem) instead of file. HTTP request body can be set from the buffer.

Reply:
`tarGz(path string, w io.Writer) error`

Reply:
what if someone accidentally has a huge file in their directory? if we do decide to tar to memory then we should have a max tar size in mind that gets enforced. I think using a temporary path is fine.

I would prefer to use a MkdirTemp: https://pkg.go.dev/os#MkdirTemp and pass empty string for dir. 
> If dir is the empty string, MkdirTemp uses the default directory for temporary files, as returned by TempDir. Multiple programs or goroutines calling MkdirTemp simultaneously will not choose the same directory. It is the caller's responsibility to remove the directory when it is no longer needed.

Reply:
Code you could use:
```
	dir, err := os.MkdirTemp("", "substratus")
	if err != nil {
		log.Fatal(err)
	}
	defer os.RemoveAll(dir) // clean up

	tarPath := filepath.Join(dir, "archive.tar.gz")
```

Reply:
I initially started down the path of writing to memory but verifying things manually was more straightforward with a local file. I think I'll go back that direction now that I've seen it work with files on disk.

Memory vs disk - I think it'd probably be a mistake if the user wanted to upload and build a container image (specifically a loader, notebook, server) and it were so big that it didn't fit into memory (eg 15Gb+). Also, an upload that large has other places it might fail - for the moment I hardcoded the expiration time on the upload URL to 300 seconds. I'll verify what exactly this applies to this in follow up, but now that we automate the client-side but if that's the window for the entire upload, really large files might fail on slow connections anyway.

I would guess expiration duration actually applies to the initiation of the HTTP PUT and not the full duration of the upload but it's worth verifying. If it refers to initiation of the HTTP request, I bet we can safely shorten the window to 10 seconds.

Reply:
I think it's fairly common in ML world to just use huge container images. That's how Substratus started out originally as well. So oI actually think using disk is quite important.

Reply:
Happy to follow a strong opinion here! 👍 tempdir incoming